tokio/src/task/local.rs

```diff
@@ -416,88 +389,58 @@ impl Scheduler {
     }
 
-    fn next_task(&self, tick: u8) -> Option<Task<Self>> {
-        if 0 == tick % CHECK_REMOTE_INTERVAL {
-            self.next_remote_task().or_else(|| self.next_local_task())
-        } else {
-            self.next_local_task().or_else(|| self.next_remote_task())
-        }
-    }
-
-    fn next_local_task(&self) -> Option<Task<Self>> {
-        unsafe { (*self.local_queue.get()).pop_front() }
-    }
-
-    fn next_remote_task(&self) -> Option<Task<Self>> {
-        // there is no semantic information in the `PoisonError`, and it
-        // doesn't implement `Debug`, but clippy thinks that it's bad to
-        // match all errors here...
-        #[allow(clippy::match_wild_err_arm)]
-        let mut lock = match self.remote_queue.lock() {
-            // If the lock is poisoned, but the thread is already panicking,
-            // avoid a double panic. This is necessary since `next_task` (which
-            // calls `next_remote_task`) can be called in the `Drop` impl.
-            Err(_) if std::thread::panicking() => return None,
-            Err(_) => panic!("mutex poisoned"),
-            Ok(lock) => lock,
-        };
-        lock.pop_front()
-    }
-
     fn tick(&self) {
         assert!(self.is_current());
         for _ in 0..MAX_TASKS_PER_TICK {
             let tick = self.tick.get().wrapping_add(1);
             self.tick.set(tick);
-            let task = match self.next_task(tick) {
+
+            let task = match unsafe {
+                // safety: we must be on the local thread to call this. The assertion
+                // the top of this method ensures that `tick` is only called locally.
+                self.queues.next_task(tick)
+            } {
                 Some(task) => task,
                 None => return,
             };
 
             if let Some(task) = task.run(&mut || Some(self.into())) {
                 unsafe {
-                    // we are on the local thread, so this is okay.
-                    self.schedule_local(task);
+                    // safety: we must be on the local thread to call this. The
+                    // the top of this method ensures that `tick` is only called locally.
+                    self.queues.push_local(task);
                 }
             }
         }
     }
-
-    fn drain_pending_drop(&self) {
-        for task in self.pending_drop.drain() {
-            unsafe {
-                (*self.tasks.get()).remove(&task);
-            }
-            drop(task);
-        }
-    }
-}
-
-impl fmt::Debug for Scheduler {
-    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
-        fmt.debug_struct("Scheduler { .. }").finish()
-    }
 }
 
 impl Drop for Scheduler {
     fn drop(&mut self) {
+        // Close the remote queue
+        self.queues.close_remote();
+
         // Drain all local tasks
-        while let Some(task) = self.next_local_task() {
+        while let Some(task) = unsafe { self.queues.next_local_task() } {
```

Comment:
It looks like draining local tasks on `Scheduler` is the same as `BasicScheduler`. Maybe a default `close_local` method could be added to `Queues` similar to `close_remote`?

Reply:
yup, I added it in a01f5963215178b2a9042b996fe99e8a18987fe3. but, I also realized that only the loop that drains remaining owned tasks is different between the local set and basic scheduler, so i put all the other fns in a single `fn shutdown`.